torchdata/nodes/_populate_queue.py

```diff
@@ -0,0 +1,40 @@
+# pyre-unsafe
+import queue
+import threading
+from typing import Iterable
+
+from torchdata.nodes.exception_wrapper import ExceptionWrapper, StartupExceptionWrapper
+
+
+def _populate_queue(
+    source: Iterable,
+    q: queue.Queue,
+    semaphore: threading.BoundedSemaphore,
+    stop_event: threading.Event,
+):
+    """Note that this is only intended to be used
+    by a single thread at once. Each instance creates its own iter for source so
+    if this is called with multiple threads, you may get duplicates if
+    source is not sharded properly.
+    """
+    try:
+        src_iter = iter(source)
+    except Exception:
+        e = StartupExceptionWrapper(where="in _populate_queue startup for device")
+        q.put(e)
+        return
+
+    while not stop_event.is_set():
+        if not semaphore.acquire(blocking=True, timeout=1.0):
```

Comment:
I'm not sure if we're gaining anything using a semaphore to control the size of the queue, rather than just using a queue with a fixed size: https://docs.python.org/3/library/queue.html#queue.Queue

It would make sense to me if we were trying to acquire the semaphore in a non-blocking manner, and then went off and did something useful when we couldn't acquire the semaphore. But we are blocking, so I think the behavior will be the same as if we had a blocking queue?

Reply:
This isn't obvious, basically the reason I went this way instead of using the sized-queue to block is that we're then holding some state in the stack implicitly, and if a user wanted to do some kind of checkpointing or state saving, this would be hard to expose to them otherwise. TBH this might not be necessary, it depends on how we're going to handle checkpointing though.

Reply:
Hmm. If we're going to checkpoint the current size of the queue, then I assume we would also want to checkpoint the contents of the queue - which would also imply its size.

In general, though, I don't think we'll want to checkpoint the internal state of the runtime system. Rather, I think we'll want to checkpoint the actual state built up as a result of processing data.

To (maybe) make this a bit more clear: let's say we're reading from a file, and every operator we're processing is stateless - it's just applying a function to a data item, producing another data item. Only the input data item effects the output data item. In such a system, our checkpoints would only need to be at the source, where we're reading the file, and at the sink, to know what we've produced on the output. Everything in between can be lost; we build it back up when the source recovers, and starts reading from the correct place in the file.

We can talk more on this later - I don't think it's critical now.